### DISCUSSION TITLE
Make Code Block more powerful

### DISCUSSION THREAD

**Discussion by Benbinbin** [facebook/lexical#5426] [Ideas]
## Highlight

The key point of Code Block is how to highlight, the `@lexical/code` module use [Prism.js](https://github.com/facebook/lexical/blob/main/packages/lexical-code/package.json#L18), but it has some [limitaions](https://prismjs.com/#limitations).

I want to know if there're other tools to make code highlight and find [code-hike/lighter](https://github.com/code-hike/lighter) is an alternative solution.

> [!TIP]
>
> [Code Hike](https://codehike.org/) built by [@pomber](https://twitter.com/pomber) is so wonderful, which is a plugin for MDX (an extentable Markdown syntax).
>
> It provides a lot of useful components to make smooth animation for "code walkthroughs", can make blog posts, documentation, coding tutorial more fun.
>
> And I also like this Twitter [thread](https://twitter.com/pomber/status/1576900746896736257) from him, showing the potential of web to show code and explanation for code block.

I also find a [plugin](https://rehype-pretty-code.netlify.app/)  for Rehype, call `rehype-pretty-code` for Rehype, 

> [!NOTE]
>
> [Rehype](https://github.com/rehypejs/rehype) is a tool that transforms HTML between ASTs (abstract syntax trees).

It can provide more powerful feature, like the **_line_** or **_word_** highlight within the code block, and **_inline code_ highlight**.

It use [shiki](https://shiki.matsu.io/) as code highlighter, which use some library, like [vscode-textmate](https://github.com/microsoft/vscode-textmate), to tokenize code text and highlight them.

> [!TIP]
>
> [Shikiji](https://shikiji.netlify.app/) is an ESM-focused rewrite of shiki, and add more features to make code [more interactive](https://shikiji.netlify.app/packages/twoslash) at web.

Although most of these resources are about Markdown or MDX, I think some idea from them is also suitable for Rich Text Editor.

**Comment by Benbinbin** [comment]
[Shikiji Transformers](https://shikiji.netlify.app/packages/transformers) can parse specific annotation into a special style, like **line highlight**, **line focus**, **whitespace and tab visual hint**, these are some great examples to make code block more powerful.

**Reply by SalahAdDin** [reply]
We have also [**HighlightJS**](https://highlightjs.org/) as an alternative.

**Comment by Benbinbin** [comment]
## Writter Experience

I think [CodeMirror](https://codemirror.net/try/) is good enough alternative as the block code for a rich text editor, it has provided more fluent UX, like **autocompletion**, **linting**

Now the Lexical CodeNode provide some basic UX, like auto indent for the insert new line, but it just [preserve the same indent as the previous line](https://github.com/facebook/lexical/blob/main/packages/lexical-code/src/CodeNode.ts#L255-L304), this is not "smart" enough

and some special behavior of block code seems to **make the Lexical Core module more complex**, for example the [insert nodes](https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalSelection.ts#L1326-L1335) 

If we want to provide the really smooth writter experience, maybe we should use some library (like the [textmate](https://github.com/microsoft/vscode-textmate) for VS Code, [Lezer](https://lezer.codemirror.net/) for CodeMirror, [Tree-sitter](https://tree-sitter.github.io/tree-sitter/) for Github) to parser the code and build the syntax tree, not just listen some COMMAND trigger by user

**Comment by yf-yang** [comment]
Thanks for starting this discussion thread! It could help more people who are interested in using code blocks.

Lexical is a rich text editor framework, which means that: 1. You can build a code editor based on Lexical. 2. Lexical is not specialized for code editor scenarios. This implies that if you want to build a code editor, Lexical offers you the flexibility to create your own; however, it does require some effort to do so. If you're looking for a solution that requires minimal effort, then it might not be as suitable as CodeMirror.

I don't believe that integrating CodeMirror into the code node is a wise idea (nor do I think the maintainers have the time or willingness to undertake such a task). In fact, if I understand correctly, the author of CodeMirror also developed another rich text editor, ProseMirror. Lexical is inspired by ProseMirror, applying React-like intrinsic mechanisms for improved functionality. From my experience, their behaviors are quite similar. Both manage an internal state, and you can configure their views. The difference is that in CodeMirror, one part is the state, and the other is the view, whereas in Lexical, states are composed of nodes, each with its own `exportDOM` method to configure the view. Therefore, you are suggesting add a state management framework into another, which is weird (CodeMirror is more like a vanilla JS framework and Lexical is more like a hybrid of React and vanilla JS).

Returning to your original question: if you want to enhance Lexical to make it more suitable for a code editor, you could create plugins for code completion, linting, etc. The idea would be to listen to editor changes, then trigger an event to communicate with your language server. When you receive information in response, update the editor to modify Lexical's internal state.

**Reply by Benbinbin** [reply]
Thanks, build a fully functional code editor from the ground up just using the Lexical sound great! 

But this way must be harder than using existing solution, like CodeMirror.

Still looking forward to a "native" and better UX block code for Lexical.

**Comment by SalahAdDin** [comment]
Is there any official consideration on this issue?
